### PR TITLE
[WB-3835][CLI-542] Keras gradient logging fix

### PR DESCRIPTION
**Pull request by farizrahman4u**

### CODE REVIEW

File: wandb/integration/keras/keras.py [hunk 1]
```diff
@@ -642,58 +704,19 @@ def _log_weights(self):
...
     def _log_gradients(self):
-        if not self.training_data:
-            raise ValueError("Need to pass in training data if logging gradients")
-
-        X_train = self.training_data[0]
-        y_train = self.training_data[1]
+        weights = self.model.trainable_weights
+        grads = [np.zeros(tuple(w.shape)) for w in weights]
+        for x, y in self._training_data_generator():
+            with tf.GradientTape() as tape:
```
Comment (raubitsj) [inline wandb/integration/keras/keras.py:710]: Awesome, how does this handle tf1.x?  can we make sure it fails (or doesnt log gradients) gracefully?

File: wandb/integration/keras/keras.py [hunk 2]
```diff
@@ -328,6 +340,51 @@ def __init__(
...
+                self.num_batches = 0
+
+            def on_batch_end(self, batch, logs=None):
+                for g, w in zip(self.grads, self.model.trainable_weights):
+                    g += w.numpy()
+                self.model.set_weights(self.og_weights)
+                self.num_batches += 1
+
+            def get_grads(self):
+                return [g / self.num_batches for g in self.grads]
```
Comment (vanpelt) [inline wandb/integration/keras/keras.py:377]: Defining the classes in this method feels dirty

File: wandb/integration/keras/keras.py [hunk 3]
```diff
@@ -642,58 +701,19 @@ def _log_weights(self):
...
-        if not self.training_data:
-            raise ValueError("Need to pass in training data if logging gradients")
-
-        X_train = self.training_data[0]
-        y_train = self.training_data[1]
+        self._grad_accumulator_model.fit(
+            self._training_data_x,
+            self._training_data_y,
+            verbose=0,
+            callbacks=[self._grad_accumulator_callback],
```
Comment (vanpelt) [inline wandb/integration/keras/keras.py:708]: Does this do an entire epoch by default?  The previous implementation just passed the data through and extracted the gradients.  I'm not super clear on what's happening with our custom optimizer but I wonder if we could get this data more efficiently.

File: wandb/integration/magic.py [hunk 4]
```diff
@@ -294,6 +294,10 @@ def _magic_fit(
     *args,
     **kwargs
 ):
+    if hasattr(self, "_wandb_internal_model"):
+        return self._fit(
+            x=x, y=y, batch_size=batch_size, epochs=epochs, *args, **kwargs
+        )
```
Comment (vanpelt) [inline wandb/integration/magic.py:300]: When would this code get executed?  We should definitely add a test to `test_keras` that assers we actually log gradients as well.

File: wandb/integration/keras/keras.py (hunk 3 above)
Comment (farizrahman4u) [reply wandb/integration/keras/keras.py:708]: That was causing OOM. Here we are basically reusing keras's fit logic (so generators, tf.Dataset etc work out of the box). Extracting gradients batch by batch manually (using `tf.GradientTape`) also causes OOM.

File: wandb/integration/magic.py (hunk 4 above)
Comment (farizrahman4u) [reply wandb/integration/magic.py:300]: This code is reached when `.fit()` is called on the grad accumulator model (if in magic mode). I have left a comment on the `_wandb_internal_model` attribute in `keras.py`

**Review by vanpelt — APPROVED** [review]
Nice
